zigzag-circuit.md

```diff
@@ -0,0 +1,168 @@
+
+#### ZigZag Overview
+
+ZigZag PoRep is based on layering DRG graphs `LAYERS` times. The data represented in each DRG layer is the data encoded in the previous layer. The final layer is the replica (which in Filecoin terms is the sealed sector).
+
+- `ReplicaId` is a unique replica identifier (see the Filecoin Proofs spec for details)
+- `CommD` is the Merkle Tree root hash of the input data to the first layer
+- `CommR[l]` is the Merkle Tree hash of the output of the DRG encoding at each layer `l` 
+- `CommRStar` is the hash of the concatenation of the `ReplicaId` and all the `CommR`s.
+
+The (offline) proof size in the ZigZag is too large for blockchain usage (~3MB). We use SNARKs to generate a proof of knowledge of a correct ZigZag proof. In other words, we implement the ZigZag proof verification algorithm in an arithmetic circuit and use SNARKs to prove that it was evaluated correctly.
+
+This circuit proves that given a Merkle root `CommD`, `CommRLast`, and `commRStar`, that the prover knew the correct replicated data at each layer.
+
+#### Spec notation
+
+- **Fr**: Field element of BLS12-381
+- **UInt**: Unsigned integer
+- **{0..x}**: From `0` (included) to `x` (not included) (e.g. `[0,x)` )
+- **Check**: 
+  - If there is an equality, create a constraint
+  - otherwise, execute the function
+- **Inclusion path**: Binary representation of the Merkle tree path that must be proven packed into a single `Fr` element.
+
+## Offline PoRep circuit
+
+#### Public Parameters
+
+*Parameters that are embeded in the circuits or used to generate the circuit*
+
+- `LAYERS : UInt`: Number of DRG layers.
+- `LAYER_CHALLENGES : [LAYERS]UInt`: Number of challenges per layer.
+- `EXPANSION_DEGREE: UInt`: Degree of each bipartite expander graph to extend dependencies between layers.
+- `BASE_DEGREE: UInt`: Degree of each Depth Robust Graph.
+- `TREE_DEPTH: UInt`: Depth of the Merkle tree. Note, this is (log_2(Size of original data in bytes/32 bytes per leaf node)).
+- `PARENT_NODES : UInt`: Defined as `EXPANSION_DEGREE+BASE_DEGREE`.
+
+#### Public Inputs
+
+*Inputs that the prover uses to generate a SNARK proof and that the verifier uses to verify it*
+
+- `ReplicaId : Fr`: A unique identifier for the replica.
+- `CommD : Fr`: the Merkle tree root hash of the original data (input to the first layer).
+- `CommRLast : Fr`: The Merkle tree root hash of the final replica (output of the last layer).
+- `CommRStar : Fr`: A commitment to each `l` layer's Merkle tree root hash `CommR[l]` and `ReplicaId`.
+- `InclusionPath : [LAYERS][]Fr`: Inclusion path for the challenged data and replica node.
+- `ParentInclusionPath : [LAYERS][][PARENT_NODES]Fr`:  Inclusion path for the parent nodes of the corresponding `InclusionPath[l][c]` nodes.
+
+##### Design notes
+
+- `CommRLast` is a public input, since we will be using it during Proof-of-Spacetime.
+- `InclusionPath` and `ParentInclusionPath`: Each layer `l` has `LAYER_CHALLENGES[l]` inclusion paths.
```

Comment:
It's a bit confusing that we explain the number of challenges per layer here, but don't mention it below. I understand why you didn't, but I think it would be worth either a slight change to notation or else an annotation in the descriptions wherever we have `[LAYERS][]…`.